### DISCUSSION TITLE
Open AI API calls

### DISCUSSION THREAD

**Discussion by frankie-mur** [kmesiab/ai-code-critic#61] [Q&A]
Taking a quick look at the recent PR #50 and through the app, I noticed the code now calls the Open AI API for each git diff file, when previously made one API call once per PR. I can see this becoming an issue with a large PR with multiple files, therefore, multiple API calls. With an emphasis on reducing token consumption overall in the app, I am hesitant if this is an ideal approach. Here is the snippet I am referring to in `main.go` 
```go
       ...
       // Step 1: Split and parse PR contents into individual GitDiff objects.
	gitDiffs := critic.ParseGitDiff(prContents, ignoreList)

	// Channel to collect responses from goroutines.
	reviewsChan := make(chan string, len(gitDiffs))
	var wg sync.WaitGroup

	for _, gitDiff := range gitDiffs {
		wg.Add(1)
		go func(gd *critic.GitDiff) {
			defer wg.Done()

			// Step 2: Send the diff to GetCodeReviewFromAPI in a goroutine.
			review, err := critic.GetCodeReviewFromAPI(gd.DiffContents, gptModel)
			if err != nil {
				reviewsChan <- fmt.Sprintf("Error getting review for %s: %s\n", gd.FilePathNew, err)
				return
			}

			// Prepend the diff header to the review.
			header := fmt.Sprintf("Review for %s:\n", gd.FilePathNew)
			reviewsChan <- header + review
		}(gitDiff)
	}
	...
```

**Comment by kmesiab** [comment]
You make some good points.  Specifically an emphasis on token reduction.   Any thoughts and work that leads to improvements on this front are like gold to the project.  :) 

**However, to alleviate some concern, I will expand on the thinking behind this decision:**

## My Initial Thoughts 

You are very right about the danger of large diffs.  This reared its head early in the project when very large `go.sum` files were included in the diffs.  Other tokens were wasted on inconsequential diffs, like diff files that only contained whitespace changes.

As a result, the ability to filter files became a necessary feature that unfortunately never got addressed until this PR.  Even now, it's implementation is hacked together and requires enhancement.  (I have delayed that enhancement in favor of first implementing a fully featured user configuration functionality).

So in the mean time, we have an environment variable to set up some filters.   Better than nothing, for now.  The net effect is that we will process much less unimportant files than before and that will reduce the token count.

Splitting the diffs and sending them individually increases the token count by the size of the prompt * the files in the diff, so it does consume extra tokens, and the trade off may not always be worth it... 🤔 


## Split Diffs 

The goal behind splitting the diffs was to enable targeted analysis in the hopes of producing higher quality reviews.   That worked, the reviews became much more rich with better quality code suggestions.   However it does also suffer from a lack of context when reviewing diffs in isolation.   It can't tell whether a change is meaningful or proper in terms of the other changes in the PR as a whole.

I have played around with the idea of instead, constructing the prompt differently.  The process would be something like:

1. Fetch the pr
3. Split the pr into individual diffs. 
4. Construct prompt:
5. Request prompt results in json
6. Process the json for better display

Ex:

```
Here is the full pr for context: 

"""" 
(Full PR here)
""""

File: a/foo.go - b/foo.go
(Diff here)

File a/bar.go - b/foo.go
(Diff here)

""""

Review each individual diff in isolation then in context.

Output the results in the following json format:

{ 
   "title": "EXAMPLE",
   "file": "NAME.ext",
   "suggestions": [
      {
         "start_line": 42,
         "end_line": 43,
         "suggestion" : "Idiomatic go suggests naming variables that are acronyms in all caps.  Consider changing ...."
      },
     {
         "start_line": 109,
         "end_line": 122,
         "suggestion" : "The change looks good overall, but you may consider adding more test coverage for cases when...."
      },
   ....
}...
...
Etc
```

This would accomplish the 1.) directed targeted individual diff review 2.) preserve total context the change lives within, and 3.) Bringing the network calls down to one and decreasing the current payload size by some reasonable factor when there are a large number of files in the PR.

## Concept

This puts us in a position where we can now, instead of producing just a rich standalone report, we can "inline" comments into the diff itself in the right hand window. Much like the rabbitai GitHub app we use. 

Much like this:

Screenshot:

<img width="819" alt="image" src="https://github.com/kmesiab/ai-code-critic/assets/161768/4c1ade36-ea41-497d-b089-a65a37e737ef">

This constituted a big change so it deserves some conversation. 

### A few final thoughts: 

1.) Since we are charged per token, the number of extra tokens added to each diff as a result of splitting the diff and sending it with its own prompt works out like this:


Imagine sending one full diff split amongst 50 files, for a tottal of 1,000 tokens, which cost $0.03. 

In the new setup, we're handling 50 individual diffs, and each has 220 extra tokens (on top of the original split tokens) from the additional prompt. 

This means every block now has 240 tokens (20 original + 220 extra). So, the cost per request has gone up to around $0.33, marking an increase of about $0.30. 

(_This calculation is based on the 8k model rate of $0.03 per 1,000 tokens._)

2. There is a trade off in here and somewhere there is a smart blance.  I think we can figure that out together. :) 

What are your thoughts?

**Reply by kmesiab** [reply]
As an aside, now that you're an official collaborator, I will make an adjustment to the repo here to add you as a codeowner, and require 1 approval before merging.   This will give us a relief valve to ensure we (and future collaborators) are on the same page with the direction we take the app. :)

**Reply by frankie-mur** [reply]
Appreciate you adding me as an official collaborator! This is my first dive into open-source, and I'm thrilled to be part of this journey. Excited to contribute to a project that I believe in! 😃

**Reply by kmesiab** [reply]
Glad to have ya!   This should, indeed, be a rewarding journey.  Should it begin to not be, we will take a step back and make it again. :)

**Comment by frankie-mur** [comment]
Wow! Appreciate the **very** detailed response @kmesiab. 

**Regarding split diffs**: I concur that incorporating context directly into the prompt is a significant enhancement, and it indeed justifies the implementation of a single network call strategy. Furthermore, the addition of functionalities such- as enabling inline comments undoubtedly elevates the user experience. This, however, does raise an important consideration about our product's unique positioning in comparison to alternatives like CodeRabbit. I am enthusiastic about collaborating on brainstorming sessions to further delineate our product's distinctive features.

**Token size**: I will preface I am not very familiar with token sizes, etc. However, your observation about the cost increase in our new diff handling setup is indeed valid and something we should seriously. To strike a balance, we could be implementing a smarter system that only applies the detailed prompt structure to diffs where it's most needed, based on certain criteria like file size or complexity. We could also consider a tiered approach, where basic diff handling remains as cost-effective as before, and the more detailed, token-intensive processing is used selectively (by the user).
